### PR TITLE
Avoid using LegacyVersion from packaging

### LINKED ISSUE #2466: `import pkg_resources` triggers a DeprecationWarning on Python 3.9
I'm pretty sure this is a bug because I don't think pkg_resources is deprecated in its entirety.

```
$ PYTHONWARNINGS=error python3 -c 'import pkg_resources'
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/packaging/version.py", line 57, in parse
    return Version(version)
  File "/usr/lib/python3.9/site-packages/packaging/version.py", line 298, in __init__
    raise InvalidVersion("Invalid version: '{0}'".format(version))
packaging.version.InvalidVersion: Invalid version: '/usr/lib/python3.9/site'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.9/site-packages/pkg_resources/__init__.py", line 3239, in <module>
    def _initialize_master_working_set():
  File "/usr/lib/python3.9/site-packages/pkg_resources/__init__.py", line 3222, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python3.9/site-packages/pkg_resources/__init__.py", line 3251, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python3.9/site-packages/pkg_resources/__init__.py", line 558, in _build_master
    ws = cls()
  File "/usr/lib/python3.9/site-packages/pkg_resources/__init__.py", line 551, in __init__
    self.add_entry(entry)
  File "/usr/lib/python3.9/site-packages/pkg_resources/__init__.py", line 607, in add_entry
    for dist in find_distributions(entry, True):
  File "/usr/lib/python3.9/site-packages/pkg_resources/__init__.py", line 2056, in find_on_path
    path_item_entries = _by_version_descending(filtered)
  File "/usr/lib/python3.9/site-packages/pkg_resources/__init__.py", line 2026, in _by_version_descending
    return sorted(names, key=_by_version, reverse=True)
  File "/usr/lib/python3.9/site-packages/pkg_resources/__init__.py", line 2024, in _by_version
    return [packaging.version.parse(part) for part in parts]
  File "/usr/lib/python3.9/site-packages/pkg_resources/__init__.py", line 2024, in <listcomp>
    return [packaging.version.parse(part) for part in parts]
  File "/usr/lib/python3.9/site-packages/packaging/version.py", line 59, in parse
    return LegacyVersion(version)
  File "/usr/lib/python3.9/site-packages/packaging/version.py", line 127, in __init__
    warnings.warn(
DeprecationWarning: Creating a LegacyVersion has been deprecated and will be removed in the next major release
```

### PR DESCRIPTION
**Pull request by yan12125**
## Summary of changes

Closes #2466
Closes https://github.com/pypa/packaging/issues/368

### Pull Request Checklist
- [x] Changes have tests
- [x] News fragment added in [`changelog.d/`].
  _(See [documentation][PR docs] for details)_

[`changelog.d/`]: https://github.com/pypa/setuptools/tree/master/changelog.d
[PR docs]:
https://setuptools.readthedocs.io/en/latest/development/developer-guide.html#making-a-pull-request

### CODE REVIEW

**Comment by yan12125** [comment]
So, test_test_command_install_requirements does not work with pip 20.3. I suspect it a pip regression. Should I wait until the pip issue is resolved or it's fine to add an workaround here first and try to solve the pip issue later?

**Comment by di** [comment]
> So, test_test_command_install_requirements does not work with pip 20.3. I suspect it a pip regression. Should I wait until the pip issue is resolved or it's fine to add an workaround here first and try to solve the pip issue later?

What exactly is the regression? That `pip` won't accept the dependency links the test was originally creating?

**Comment by yan12125** [comment]
> What exactly is the regression? That pip won't accept the dependency links the test was originally creating?

pip no longer accepts local paths (e.g., `/tmp/foo/bar-1.0.tar.gz`) as dependency links. Fortunately I managed to make tests pass by turning local paths into URLs (e.g., `file:///tmp/foo/bar-1.0.tar.gz`) in the second commit 12fe738fd0b78f42a12d3aed4ee9c2863b8c31e7, so the test suite can work with pip 20.3+ now.

**Review by di — APPROVED** [review]

**Comment by yan12125** [comment]
Thanks for the review. I added news fragments for fixes in this PR.

**Review by jaraco — CHANGES_REQUESTED** [review]
Thanks for the contrib. I'm sorry this code is so bad. Your patch is looking good, although I think there are some cases not considered here and not covered by the tests. Can you add to the tests (possibly in the doctests) the cases that are being affected by this change, namely, 

- a name that was previously supported with a legacy version in the name (e.g. `1.3_04`)
- two names, one of which is legacy and the other which is pep 440
- two legacy versions

I suspect there's another bug lingering here, revealed in the bug report, where it seems that thus function is expecting a single filename, but is getting full path names. With the lenient LegacyVersion, that seemed to work adequately, but now that this code is more strict, it's not clear what should happen. It might be worth refining that expectation as well (disallow path names or document and test the expected behavior).

**Comment by jaraco** [comment]
I didn't want to wait on this issue to resolve #2485, so I cherry-picked your commit onto main (thanks for keeping them separate) and force-pushed this branch without that commit.

File: pkg_resources/__init__.py [hunk 1]
```diff
@@ -2021,7 +2021,17 @@ def _by_version(name):
         """
         name, ext = os.path.splitext(name)
         parts = itertools.chain(name.split('-'), [ext])
-        return [packaging.version.parse(part) for part in parts]
```
[the remark this answers is not in the document]
Comment (yan12125) [reply pkg_resources/__init__.py:2024]: Thanks this is indeed a breaking change. I changed the category in the latest version of the patch.

File: pkg_resources/__init__.py [hunk 2]
```diff
@@ -2021,7 +2021,17 @@ def _by_version(name):
...
+
+        def parse_version_ignore_invalid(version):
+            try:
+                return packaging.version.Version(version)
+            except packaging.version.InvalidVersion:
+                # Per PEP 440, version strings are compared in a
+                # case-insensitive manner. Do the same for
+                # non-version strings.
+                return version.lower()
```
[the remark this answers is not in the document]
Comment (yan12125) [reply pkg_resources/__init__.py]: Thanks I added a helper class to allow such comparisons and added relevant doctests.

**Comment by yan12125** [comment]
> I didn't want to wait on this issue to resolve #2485, so I cherry-picked your commit onto main (thanks for keeping them separate) and force-pushed this branch without that commit.

Thanks for cherry-picking. Tests are red again - not fully compatible with pytest 6.2+ this time. I will create another PR for that.

File: pkg_resources/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaraco) [reply pkg_resources/__init__.py:2024]: This now looks to me like we're getting dangerously close to re-implementing `LegacyVersion` as `_NonVersion`. It seems to me if packaging is removing support for the Legacy Version, then probably Setuptools should do so too. Here's what I'd like to happen:

- in the Packaging project (or possibly on another forum like packaging-problems or distutils-sig), discuss the deprecation and removal of LegacyVersion and what they intend to happen in cases where packages still use those versions.
- Follow the spirit of that change. For example, if they're just testing the waters to see who complains about the deprecation, we should complain and explain this use case that's broken by the change. If they have a plan to wean users off of LegacyVersions by dropping support for them, then we should honor that intent and wean our users off of them as well.

I don't want setuptools to be the dumping ground for legacy behavior. If users want that behavior but the PyPA wishes to dissuade that behavior, Setuptools would like to honor that intent.

[the remark this answers is not in the document]
Comment (jaraco) [reply pkg_resources/__init__.py:2024]: I left https://github.com/pypa/packaging/issues/368#issuecomment-748630684 to inquire about the intentions for packaging in the deprecation.

[the remark this answers is not in the document]
Comment (yan12125) [reply pkg_resources/__init__.py:2024]: Thanks very much for explanations. Sounds like more discussions are needed before doing code changes. I'll mark this PR as draft.

**Comment by worstpractice** [comment]
@yan12125 Great work!

Any blockers remaining for this PR?

**Comment by yan12125** [comment]
Thanks! However, I consider this pull request abandoned, as my approach will likely introduce more issues in the long term. I will close this PR in the hopes of avoiding further confusion. Please follow https://github.com/pypa/setuptools/issues/2497 instead.

**Pull request closed without merging**